src/tpm2-tss-engine-rsa.c

```diff
@@ -246,8 +261,19 @@ rsa_priv_dec(int flen,
         goto error;
     }
 
-    r = init_tpm_key(&ectx, &keyHandle, tpm2Data);
-    ERRchktss(rsa_priv_dec, r, goto out);
+    if (tpm2Data->privatetype == handle) {
+      r = Esys_Initialize(&ectx, NULL, NULL);
+      ERRchktss(rsa_priv_dec, r, goto out);
+      r = Esys_Startup(ectx, TPM2_SU_CLEAR);
+      ERRchktss(rsa_priv_dec, r, goto out);
+      r = Esys_TR_FromTPMPublic(ectx, tpm2Data->handle, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &keyHandle);
+      ERRchktss(rsa_priv_dec, r, goto out);
+      r = Esys_TR_SetAuth(ectx, ESYS_TR_RH_OWNER, &tpm2Data->userauth);
```

Comment:
You want to do a ReadPublic call here and check the NODA flag, then if the NODA flag isn't set, use a UI to prompt for a password if you don't already have one.

See http://git.infradead.org/users/dwmw2/openconnect.git/blob/tpm2:/gnutls_tpm2_esys.c#l277 for example.

Reply:
Okay I corrected it